lib/src/device_authorization_grant.dart

```diff
@@ -0,0 +1,401 @@
+  ///
+  /// The scope strings will be separated by the provided [delimiter]. This
+  /// defaults to `" "`, the OAuth2 standard, but some APIs (such as Facebook's)
+  /// use non-standard delimiters.
+  ///
+  /// By default, this follows the OAuth2 spec and requires the server's
+  /// responses to be in JSON format. However, some servers return non-standard
+  /// response formats, which can be parsed using the [getParameters] function.
+  ///
+  /// This function is passed the `Content-Type` header of the response as well
+  /// as its body as a UTF-8-decoded string. It should return a map in the same
+  /// format as the [standard JSON response][].
+  ///
+  /// [standard JSON response]: https://tools.ietf.org/html/rfc6749#section-5.1
+  DeviceAuthorizationGrant(
+    this.identifier,
+    this.deviceEndpoint,
+    this.tokenEndpoint, {
+    this.secret,
+    String? delimiter,
+    bool basicAuth = true,
+    http.Client? httpClient,
+    CredentialsRefreshedCallback? onCredentialsRefreshed,
+    Map<String, dynamic> Function(MediaType? contentType, String body)?
+        getParameters,
+  })  : _basicAuth = basicAuth,
+        _httpClient = httpClient ?? http.Client(),
+        _delimiter = delimiter ?? ' ',
+        _getParameters = getParameters ?? parseJsonParameters,
+        _onCredentialsRefreshed = onCredentialsRefreshed;
+
+  /// starts the Flow and returns [DeviceProperties]
+  /// [useSecret] whether to use the clientSecret for the token-request or not
+  Future<DeviceProperties> getDeviceCode(
+      {Iterable<String>? scopes, bool useSecret = false}) async {
+    if (_state != _State.initial) {
+      throw StateError('The device_code has already been generated.');
+    }
+
+    var scopeList = scopes?.toList() ?? <String>[];
+    _scopes = scopeList;
+
+    var headers = <String, String>{};
+    var body = <String, String>{};
+
+    var secret = this.secret;
+    if (_basicAuth && secret != null && useSecret) {
+      headers['Authorization'] = basicAuthHeader(identifier, secret);
+    } else {
+      // The ID is required for this request any time basic auth isn't being
+      // used, even if there's no actual client authentication to be done.
+      body['client_id'] = identifier;
+      if (secret != null && useSecret) body['client_secret'] = secret;
+    }
+
+    if (scopeList.isNotEmpty) body['scope'] = scopeList.join(_delimiter);
+
+    var startTime = DateTime.now();
+
+    var response =
+        await _httpClient.post(deviceEndpoint, headers: headers, body: body);
+
+    var device_properties = _handleDeviceCodeResponse(
+        response, startTime, _scopes,
+        getParameters: _getParameters);
+
+    _state = _State.polling;
+
+    return device_properties;
+  }
+
+  /// Check if the user has confirmed the device and then return a [Client]
+  /// [useSecret] whether to use the clientSecret for the token-request or not
+  Future<Client> pollForToken({bool useSecret = true}) async {
```

Comment:
How does this work if the user haven't signed in yet?

Will it throw an `Exception`? If so we should document what it throws?

Reply:
Perhaps, it's better to return `null` or throw a very specific exception like `DeviceAuthorizationPendingException` ?

Or maybe we should just do: `grant.pollForToken({bool useSecret = true, Duration timeout = const Duration(Minutes: 5)})`, and then have the function repeatedly polling the server for up to 5 minutes (or whatever default we think is reasonable). That way the user of this library won't have to implement a polling loop themselves.

And a user who wants to implement their own polling loop can just do `grant.pollForToken(timeout: Duration())`.

Polling with `device_code.interval ?? 10` seems very reasonable, and if we implement the polling logic here, odds are that users are more likely implement the logic correctly :D

I'm okay, with not putting the polling logic in here too... WDYT?